language/gpt3/megatron/dataset.py

```diff
@@ -0,0 +1,171 @@
+        args.vocab_extra_ids = 0
+        args.tokenizer_model = "./data/c4_en_301_5Mexp2_spm.model"
+        self.tokenizer = build_tokenizer(args)
+
+        self.list_data_dict = utils.jload(self.dataset_path)
+
+        prompt_input, prompt_no_input = (
+            PROMPT_DICT["prompt_input"],
+            PROMPT_DICT["prompt_no_input"],
+        )
+        self.sources = [
+            prompt_input.format_map(example) for example in self.list_data_dict
+        ]
+        self.targets = [f"{example['output']}" for example in self.list_data_dict]
+
+        (
+            self.source_encoded_input_ids,
+            # self.source_encoded_attn_masks,
+            self.source_encoded_input_id_lengths,
+        ) = self.encode_samples()
+        self.count = total_count_override or len(self.sources)
+        self.perf_count = perf_count_override or self.count
+
+    def _build_attention_mask(self, tokens):
+        """Build the attention mask and postition ids for the input tokens."""
+
+        # Since we are not interested in loss-mask and reset attention/position
+        # is also False, eod_token is not used so it is safe to set it to None.
+        attention_mask, _, position_ids = get_ltor_masks_and_position_ids(
+            data=tokens,
+            eod_token=None,
+            reset_position_ids=False,
+            reset_attention_mask=False,
+            eod_mask_loss=False,
+        )
+        return attention_mask
+
+    def tokenize_prompts(self, prompts, tokens_to_generate, add_BOS):
+        """Given a set of prompts and number of tokens to generate:
+        - tokenize prompts
+        - set the sequence length to be the max of length of prompts
+        plus the number of tokens we would like to generate
+        - pad all the sequences to this length so we can convert them
+        into a 2D tensor.
+        """
+
+        # Tokenize all the prompts.
+        if add_BOS:
+            prompts_tokens = [
+                [self.tokenizer.eod] + self.tokenizer.tokenize(prompt)
+                for prompt in prompts
+            ]
+        else:
+            prompts_tokens = [self.tokenizer.tokenize(prompt) for prompt in prompts]
+
+        # Now we have a list of list of tokens which each list has a different
+        # size. We want to extend this list to:
+        #   - incorporate the tokens that need to be generated
+        #   - make all the sequences equal length.
+        # Get the prompts length.
+        prompts_length = [len(prompt_tokens) for prompt_tokens in prompts_tokens]
+        # Get the max prompts length.
+        max_prompt_len = max(prompts_length)
+        # Number of tokens in the each sample of the batch.
+        samples_length = max_prompt_len + tokens_to_generate
+        # Now update the list of list to be of the same size: samples_length.
+        for prompt_tokens, prompt_length in zip(prompts_tokens, prompts_length):
+            padding_size = samples_length - prompt_length
```

Comment:
Can you explain why the padding_size is `samples_length - prompt_length` instead of `max_prompt_len - prompt_length`? For example if max_prompt_len is 2048, and tokens_to_generate is 128, all inputs should be padded to 2048 if I am understanding correctly.

Reply:
If I understand the padding is just the maximum number of tokens to generate for each sample. This function is called for each sample individually, so `prompt_length = max_prompt_len` which implies `padding_size = tokens_to_generate`.

I took a look at the megatron implementation and this is the input format for the text generation functions. For example:
- Here the prompts tokenized with the argument tokens_to_generate: https://github.com/mlcommons/training/blob/64b14a9abc74e08779a175abca7d291f8c957632/large_language_model/megatron-lm/megatron/text_generation/api.py#L132-L133
- Here they are padded: https://github.com/mlcommons/training/blob/64b14a9abc74e08779a175abca7d291f8c957632/large_language_model/megatron-lm/megatron/text_generation/tokenization.py#L123-L125

Reply:
I see, this is interesting. Thanks for explaining.